draft-ietf-tls-esni.md

```diff
@@ -347,148 +408,92 @@ computed from Z as follows:
-is done to prevent nonce re-use since the client's ESNI key share, and
-thus the value of Zx, does not change across ClientHello retries.)
-
-Note that ESNIContents will not be directly transmitted to the server in the
-ClientHello. The server will instead reconstruct the same object by obtaining
-its values from ClientEncryptedSNI and ClientHello.
+Where the Hash for HKDF is the hash function associated with the HKDF
+instantiation. The salt argument for HKDF-Extract is a string
+consisting of Hash.length bytes set to zeros. For a client's first
+ClientHello, KeyLabel = "esni key" and IVLabel = "esni iv", whereas
+for a client's second ClientHello, sent in response to a
+HelloRetryRequest, KeyLabel = "hrr esni key" and IVLabel = "hrr esni
+iv". (This label variance is done to prevent nonce re-use since the
+client's ESNI key share, and thus the value of Zx, does not change
+across ClientHello retries.)
 
 [[TODO: label swapping fixes a bug in the spec, though this may not be
 the best way to deal with HRR. See https://github.com/tlswg/draft-ietf-tls-esni/issues/121
 and https://github.com/tlswg/draft-ietf-tls-esni/pull/170 for more details.]]
 
-~~~
-   struct {
-       opaque record_digest<0..2^16-1>;
-       KeyShareEntry esni_key_share;
-       Random client_hello_random;
-   } ESNIContents;
-~~~
-
-record_digest
-: Same value as ClientEncryptedSNI.record_digest.
-
-esni_key_share
-: Same value as ClientEncryptedSNI.key_share.
+The client MAY replace any large, duplicated, extensions in ClientHelloInner
+with the corresponding "outer_extensions" extension, as described in
+{{outer-extensions}}.
 
-client_hello_random
-: Same nonce as ClientHello.random.
-
-The client then creates a ClientESNIInner structure:
+The encrypted ClientHello value is then computed as:
 
 ~~~~
-   struct {
-       opaque dns_name<1..2^16-1>;
-       opaque zeros[ESNIConfig.padded_length - length(dns_name)];
-   } PaddedServerNameList;
-
-   struct {
-       uint8 nonce[16];
-       PaddedServerNameList realSNI;
-   } ClientESNIInner;
+    encrypted_sni = AEAD-Encrypt(key, iv, "", ClientHelloIInner)
 ~~~~
 
-nonce
-: A random 16-octet value to be echoed by the server in the
-"encrypted_server_name" extension.
+[[OPEN ISSUE: If in the future you were to reuse these keys for
+0-RTT priming, then you would have to worry about potentially
+expanding twice of Zx We should think about how
+to harmonize these to make sure that we maintain key separation.]]
 
-dns_name
-: The true SNI DNS name, that is, the HostName value that would have been sent in the
-plaintext "server_name" extension. (NameType values other than "host_name" are
-unsupported since SNI extensibility failed {{SNIExtensibilityFailed}}).
+Finally, the client MUST generate a ClientHelloOuter message
+containing the "encrypted_client_hello" extension with the values as
+indicated above. The client MUST place the value of
+ESNIConfig.public_name in the "server_name" extension. The remaining
```

Comment:
The draft -02 protocol seems to work without an unencrypted SNI. Is this adding a required unencrypted SNI?